rest/route/host_agent_test.go

```diff
@@ -857,295 +858,656 @@ func TestTaskLifecycleEndpoints(t *testing.T) {
+			task, shouldTeardown, err = assignNextAvailableTask(ctx, env, d.Tq1, model.NewTaskDispatchService(time.Minute), d.Host2, details)
+			require.NoError(t, err)
+			require.NotNil(t, task)
+			assert.False(t, shouldTeardown)
+			assert.Equal(t, nextTaskId, task.Id)
+
+			tq, err = model.LoadTaskQueue(d.Distro1.Id)
+			require.NoError(t, err)
+			assert.Equal(t, 1, tq.Length())
+
+			h, err = host.FindOneId(ctx, d.Host2.Id)
+			require.NoError(t, err)
+			assert.Equal(t, nextTaskId, h.RunningTask)
+
+			// The third host should not get a task, since the limit is 2.
+			task, shouldTeardown, err = assignNextAvailableTask(ctx, env, d.Tq1, model.NewTaskDispatchService(time.Minute), extraHost, details)
+			fmt.Print(task, shouldTeardown, err)
+			require.NoError(t, err)
+			assert.Nil(t, task)
+			assert.False(t, shouldTeardown)
+
+			tq, err = model.LoadTaskQueue(d.Distro1.Id)
+			require.NoError(t, err)
+			assert.Equal(t, 0, tq.Length())
+
+			h, err = host.FindOneId(ctx, extraHost.Id)
+			require.NoError(t, err)
+			assert.Equal(t, "", h.RunningTask)
+		},
+	} {
+		t.Run(fmt.Sprintf("%s with %s dispatcher settings", tName, evergreen.DispatcherVersionRevisedWithDependencies), func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			env := &mock.Environment{}
+			require.NoError(t, env.Configure(ctx))
+
+			colls := []string{distro.Collection, host.Collection, task.Collection, model.TaskQueuesCollection, model.ProjectRefCollection, model.VersionCollection, build.Collection}
+			require.NoError(t, db.ClearCollections(colls...))
+			require.NoError(t, modelUtil.AddTestIndexes(host.Collection, true, true, host.RunningTaskKey))
+
+			var data data
+
+			data.Project1 = &model.ProjectRef{
+				Id:      "exists",
+				Enabled: true,
+			}
+			require.NoError(t, data.Project1.Insert())
+			data.Project2 = &model.ProjectRef{
+				Id:      "also-exists",
+				Enabled: true,
+			}
+			require.NoError(t, data.Project2.Insert())
+
+			data.Distro1 = &distro.Distro{
+				Id: "d1",
+				DispatcherSettings: distro.DispatcherSettings{
+					Version: evergreen.DispatcherVersionRevisedWithDependencies,
 				},
+			}
+			require.NoError(t, data.Distro1.Insert(ctx))
+			data.Host1 = &host.Host{
+				Id:     "h1",
+				Distro: *data.Distro1,
 				Secret: hostSecret,
+				Status: evergreen.HostRunning,
 			}
-			So(anotherHost.Insert(ctx), ShouldBeNil)
-			h2 := host.Host{
-				Id: "host2",
-				Distro: distro.Distro{
-					Id: d.Id,
+			require.NoError(t, data.Host1.Insert(ctx))
+			data.Host2 = &host.Host{
+				Id:     "h2",
+				Distro: *data.Distro1,
+				Secret: hostSecret,
+				Status: evergreen.HostRunning,
+			}
+			require.NoError(t, data.Host2.Insert(ctx))
+			data.Version1 = &model.Version{Id: "v1"}
+			require.NoError(t, data.Version1.Insert())
+			data.BuildVariant1 = &build.Build{Id: "bv1", BuildVariant: "bv1", Version: data.Version1.Id, Tasks: []build.TaskCache{{Id: "tg1-task1"}, {Id: "tg1-task2"}}}
```

Comment:
tiny nit: it's easier to read structs later when they're broken into lines (like host below) if there's more than one field being set

Reply:
Made it multiline!